Import fridge-tag not working

## What went wrong? 😲

When I tried to import Fridge-Tag in 'monitoring', it displays an error message and do not show the datas.
I tried with Fridge 2E and 2L.

<img width="1254" alt="Image" src="https://github.com/user-attachments/assets/5c30c08a-0a99-495a-8ebd-1ba1efcd98c6" />

## Expected behaviour 🤔

## How to Reproduce 🔨

Steps to reproduce the behaviour:

1. Go to 'Monitoring'
2. Click on 'Import Fridge-Tag'
3. Choose txt. file
4. See error

## Your environment 🌱

<!-- e.g. 1.2.3 -->

- Open mSupply Version: demo 2.7.00-RC5
- Legacy mSupply Central Server Version: 
<!-- e.g. android, browser (extra points if you tell us which one), desktop (windows), desktop (macOS), server (windows) -->
- Platform: Mac
<!-- PostgreSQL or SQLite3 -->
- Database type:

This is on DJI demo server, looking at the logs i find this:

2025-04-23 06:59:14.305872700 [ERROR] <server::upload_fridge_tag:52>:Error while integrating sensor data -> [
    "Cannot convert to local timestamp",
]

When uploading

--- a/server/service/src/sensor/berlinger.rs
+++ b/server/service/src/sensor/berlinger.rs
@@ -299,69 +299,80 @@ fn convert_from_localtime(
                      }| {
                         let local = match Local.from_local_datetime(&timestamp) {
                             LocalResult::None => {
-                                return Err(anyhow::anyhow!("Cannot convert to local timestamp"))
+                                log::error!("Cannot convert to local timestamp {}", timestamp);
+                                timestamp
                             }
-                            LocalResult::Single(r) => r,
-                            LocalResult::Ambiguous(r, _) => r,
+                            LocalResult::Single(r) => r.naive_utc(),
+                            LocalResult::Ambiguous(r, _) => r.naive_utc(),
                         };
-                        Ok(temperature_sensor::TemperatureLog {
+                        temperature_sensor::TemperatureLog {
                             temperature,
-                            timestamp: local.naive_utc(),
-                        })
+                            timestamp: local,
+                        }
                     },
                 )
-                .collect::<Result<_, _>>()?,
+                .collect::<Vec<_>>(),
         ),
     };
     // map temperature breaches
-    let breaches_mapped: Option<Vec<temperature_sensor::TemperatureBreach>> = match sensor
-        .clone()
-        .breaches
-    {
-        None => None,
-        Some(breaches) => Some(
-            breaches
-                .into_iter()
-                .map(
-                    |temperature_sensor::TemperatureBreach {
-                         breach_type,
-                         start_timestamp,
-                         end_timestamp,
-                         duration,
-                         acknowledged,
-                     }| {
-                        let local_start = match Local.from_local_datetime(&start_timestamp) {
-                            LocalResult::None => {
-                                return Err(anyhow::anyhow!("Cannot convert to local timestamp"))
+    let breaches_mapped: Option<Vec<temperature_sensor::TemperatureBreach>> =
+        match sensor.clone().breaches {
+            None => None,
+            Some(breaches) => Some(
+                breaches
+                    .into_iter()
+                    .map(
+                        |temperature_sensor::TemperatureBreach {
+                             breach_type,
+                             start_timestamp,
+                             end_timestamp,
+                             duration,
+                             acknowledged,
+                         }| {
+                            let local_start = match Local.from_local_datetime(&start_timestamp) {
+                                LocalResult::None => {
+                                    log::error!(
+                                        "Cannot convert to local timestamp (breach start) {}",
+                                        start_timestamp
+                                    );
+                                    start_timestamp
+                                }
+                                LocalResult::Single(r) => r.naive_utc(),
+                                LocalResult::Ambiguous(r, _) => r.naive_utc(),
+                            };
+                            let local_end = match Local.from_local_datetime(&end_timestamp) {
+                                LocalResult::None => {
+                                    log::error!(
+                                        "Cannot convert to local timestamp (breach end) {}",
+                                        end_timestamp
+                                    );
+                                    end_timestamp
+                                }
+                                LocalResult::Single(r) => r.naive_utc(),
+                                LocalResult::Ambiguous(r, _) => r.naive_utc(),
+                            };
+                            temperature_sensor::TemperatureBreach {
+                                breach_type,
+                                start_timestamp: local_start,
+                                end_timestamp: local_end,
+                                duration,
+                                acknowledged,
                             }
-                            LocalResult::Single(r) => r,
-                            LocalResult::Ambiguous(r, _) => r,
-                        };
-                        let local_end = match Local.from_local_datetime(&end_timestamp) {
-                            LocalResult::None => {
-                                return Err(anyhow::anyhow!("Cannot convert to local timestamp"))
-                            }
-                            LocalResult::Single(r) => r,
-                            LocalResult::Ambiguous(r, _) => r,
-                        };
-                        Ok(temperature_sensor::TemperatureBreach {
-                            breach_type,
-                            start_timestamp: local_start.naive_utc(),
-                            end_timestamp: local_end.naive_utc(),
-                            duration,
-                            acknowledged,
-                        })
-                    },
-                )
-                .collect::<Result<_, _>>()?,
-        ),
-    };
+                        },
+                    )
+                    .collect::<Vec<_>>(),
+            ),
+        };
     // convert last connected timestamp
     let last_connected_timestamp_converted = match sensor.clone().last_connected_timestamp {
         None => None,
         Some(timestamp) => Some(match Local.from_local_datetime(&timestamp) {
             LocalResult::None => {
-                return Err(anyhow::anyhow!("Cannot convert to local timestamp").into())
+                log::error!(
+                    "Cannot convert to local timestamp (last connected time) {}",
+                    timestamp
+                );
+                timestamp
             }
             LocalResult::Single(r) => r.naive_utc(),
             LocalResult::Ambiguous(r, _) => r.naive_utc(),

--- a/server/service/src/sensor/berlinger.rs
+++ b/server/service/src/sensor/berlinger.rs
@@ -284,6 +284,23 @@ pub enum ReadSensorError {
     Other(#[from] anyhow::Error),
 }
 
+fn resolve_localtime(timestamp: &NaiveDateTime, context: &str) -> Option<NaiveDateTime> {
+    // Local time conversion can fail if a time doesn't exist in the local timezone (e.g. due to Daylight Saving skipping an hour)
+    // We also need to handle the case where the time is ambiguous e.g. there can be 2 different 2am times in the same day due to daylight saving
+    match Local.from_local_datetime(timestamp) {
+        LocalResult::None => {
+            log::error!(
+                "Cannot convert to local timestamp ({}) - {}",
+                context,
+                timestamp
+            );
+            None
+        }
+        LocalResult::Single(r) => Some(r.naive_utc()),
+        LocalResult::Ambiguous(r, _) => Some(r.naive_utc()),
+    }
+}
+
 fn convert_from_localtime(
     sensor: &temperature_sensor::Sensor,
 ) -> Result<temperature_sensor::Sensor, ReadSensorError> {
@@ -292,19 +309,14 @@ fn convert_from_localtime(
         None => None,
         Some(logs) => Some(
             logs.into_iter()
+                .filter(|log| resolve_localtime(&log.timestamp, "Temperature log").is_some())
                 .map(
                     |temperature_sensor::TemperatureLog {
                          timestamp,
                          temperature,
                      }| {
-                        let local = match Local.from_local_datetime(&timestamp) {
-                            LocalResult::None => {
-                                log::error!("Cannot convert to local timestamp {}", timestamp);
-                                timestamp
-                            }
-                            LocalResult::Single(r) => r.naive_utc(),
-                            LocalResult::Ambiguous(r, _) => r.naive_utc(),
-                        };
+                        let local =
+                            resolve_localtime(&timestamp, "Temperature log").unwrap_or(timestamp);
                         temperature_sensor::TemperatureLog {
                             temperature,
                             timestamp: local,
@@ -329,28 +341,10 @@ fn convert_from_localtime(
                              duration,
                              acknowledged,
                          }| {
-                            let local_start = match Local.from_local_datetime(&start_timestamp) {
-                                LocalResult::None => {
-                                    log::error!(
-                                        "Cannot convert to local timestamp (breach start) {}",
-                                        start_timestamp
-                                    );
-                                    start_timestamp
-                                }
-                                LocalResult::Single(r) => r.naive_utc(),
-                                LocalResult::Ambiguous(r, _) => r.naive_utc(),
-                            };
-                            let local_end = match Local.from_local_datetime(&end_timestamp) {
-                                LocalResult::None => {
-                                    log::error!(
-                                        "Cannot convert to local timestamp (breach end) {}",
-                                        end_timestamp
-                                    );
-                                    end_timestamp
-                                }
-                                LocalResult::Single(r) => r.naive_utc(),
-                                LocalResult::Ambiguous(r, _) => r.naive_utc(),
-                            };
+                            let local_start = resolve_localtime(&start_timestamp, "Breach start")
+                                .unwrap_or(start_timestamp);
+                            let local_end = resolve_localtime(&end_timestamp, "Breach end")
+                                .unwrap_or(end_timestamp);
                             temperature_sensor::TemperatureBreach {
                                 breach_type,
                                 start_timestamp: local_start,
@@ -364,19 +358,11 @@ fn convert_from_localtime(
             ),
         };
     // convert last connected timestamp
-    let last_connected_timestamp_converted = match sensor.clone().last_connected_timestamp {
+    let last_connected_timestamp_converted = match sensor.last_connected_timestamp {
         None => None,
-        Some(timestamp) => Some(match Local.from_local_datetime(&timestamp) {
-            LocalResult::None => {
-                log::error!(
-                    "Cannot convert to local timestamp (last connected time) {}",
-                    timestamp
-                );
-                timestamp
-            }
-            LocalResult::Single(r) => r.naive_utc(),
-            LocalResult::Ambiguous(r, _) => r.naive_utc(),
-        }),
+        Some(last_connected_timestamp) => {
+            resolve_localtime(&last_connected_timestamp, "Last connected timestamp")
+        }
     };
 
     let mut sensor_mapped = sensor.clone();

